File: C#/Silverlight/source/LSilverlight-0.3.3/Loon.Core.Graphics.Component/BSPCollisionNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Loon.Core.Geom;
using Loon.Utils;
using Loon.Java.Collections;
using System.Collections;

namespace Loon.Core.Graphics.Component
{
    public sealed class BSPCollisionNode
    {

        private Dictionary<Actor, ActorNode> actors;

        private BSPCollisionNode parent;

        private RectBox area;

        private float splitAxis;

        private float splitPos;

        private BSPCollisionNode left;

        private BSPCollisionNode right;

        private bool areaRipple;

        public BSPCollisionNode(RectBox area_0, int splitAxis_1, int splitPos_2)
        {
            
            this.area = area_0;
            this.splitAxis = splitAxis_1;
            this.splitPos = splitPos_2;
            this.actors = new Dictionary<Actor, ActorNode>();
        }

        public void SetChild(int side, BSPCollisionNode child)
        {
            if (side == 0)
            {
                this.left = child;
                if (child != null)
                {
                    child.parent = this;
                }
            }
            else
            {
                this.right = child;
                if (child != null)
                {
                    child.parent = this;
                }
            }
        }

        public void Clear()
        {
            foreach (ActorNode node in actors.Values)
            {
                if (node != null)
                {
                    node.Dispose();
                }
            }
            actors.Clear();
        }

        public void SetArea(RectBox area_0)
        {
            this.area = area_0;
            this.areaRipple = true;
        }

        public void SetSplitAxis(float axis)
        {
            if (axis != this.splitAxis)
            {
                this.splitAxis = axis;
                this.areaRipple = true;
            }
        }

        public void SetSplitPos(float pos)
        {
            if (pos != this.splitPos)
            {
                this.splitPos = pos;
                this.areaRipple = true;
            }

        }

        public float GetSplitAxis()
        {
            return this.splitAxis;
        }

        public float GetSplitPos()
        {
            return this.splitPos;
        }

        public RectBox GetLeftArea()
        {
            return (this.splitAxis == 0) ? new RectBox(this.area.GetX(),
                    this.area.GetY(), this.splitPos - this.area.GetX(),
                    this.area.GetHeight()) : new RectBox(this.area.GetX(),
                    this.area.GetY(), this.area.GetWidth(), this.splitPos
                            - this.area.GetY());
        }

        public RectBox GetRightArea()
        {
            return (this.splitAxis == 0) ? new RectBox(this.splitPos,
                    this.area.GetY(), this.area.GetRight() - this.splitPos,
                    this.area.GetHeight()) : new RectBox(this.area.GetX(),
                    this.splitPos, this.area.GetWidth(), this.area.GetBottom()
                            - this.splitPos);
        }

        public RectBox GetArea()
        {
            return this.area;
        }

        private void ResizeChildren()
        {
            if (this.left != null)
            {
                this.left.SetArea(this.GetLeftArea());
            }

            if (this.right != null)
            {
                this.right.SetArea(this.GetRightArea());
            }

        }

        public BSPCollisionNode GetLeft()
        {
            if (this.areaRipple)
            {
                this.ResizeChildren();
                this.areaRipple = false;
            }
            return this.left;
        }

        public BSPCollisionNode GetRight()
        {
            if (this.areaRipple)
            {
                this.ResizeChildren();
                this.areaRipple = false;
            }
            return this.right;
        }

        public BSPCollisionNode GetParent()
        {
            return this.parent;
        }

        public void SetParent(BSPCollisionNode parent_0)
        {
            this.parent = parent_0;
        }

        public int GetChildSide(BSPCollisionNode child)
        {
            return (this.left == child) ? 0 : 1;
        }

        public void AddActor(Actor actor)
        {
            CollectionUtils.Put(actors,actor, new ActorNode(actor, this));
        }

        public bool ContainsActor(Actor actor)
        {
            ActorNode anode = (ActorNode)CollectionUtils.Get(actors,actor);
            if (anode != null)
            {
                anode.Mark();
                return true;
            }
            else
            {
                return false;
            }
        }

        public void ActorRemoved(Actor actor)
        {
            CollectionUtils.Remove(actors,actor);
        }

        public int NumberActors()
        {
            return this.actors.Count;
        }

        public bool IsEmpty()
        {
            return actors==null||this.actors.Count == 0;
        }

        public IIterator GetEntriesIterator()
        {
            return new IteratorAdapter(this.actors.Values.GetEnumerator());
        }

        public IIterator GetActorsIterator()
        {
            return new IteratorAdapter(this.actors.Keys.GetEnumerator());
        }

        public IList GetActorsList()
        {
            List<Actor> result = new List<Actor>();
            foreach (Actor a in actors.Keys)
            {
                result.Add(a);
            }
            return result;
        }

    }
}

File: Java/Android/OpenGLES/source/core/src/loon/core/graphics/component/BSPCollisionNode.java
package loon.core.graphics.component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;

import loon.core.geom.RectBox;


/**
 * 
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */

@SuppressWarnings({ "unchecked", "rawtypes" })
public final class BSPCollisionNode  {

	private HashMap<Actor, ActorNode> actors;

	private BSPCollisionNode parent;

	private RectBox area;

	private float splitAxis;

	private float splitPos;

	private BSPCollisionNode left;

	private BSPCollisionNode right;

	private boolean areaRipple;

	public BSPCollisionNode(RectBox area, int splitAxis, int splitPos) {
		this.area = area;
		this.splitAxis = splitAxis;
		this.splitPos = splitPos;
		this.actors = new HashMap<Actor, ActorNode>();
	}

	public void setChild(int side, BSPCollisionNode child) {
		if (side == 0) {
			this.left = child;
			if (child != null) {
				child.parent = this;
			}
		} else {
			this.right = child;
			if (child != null) {
				child.parent = this;
			}
		}
	}

	public void clear() {
		for (ActorNode node : actors.values()) {
			if (node != null) {
				node.dispose();
				node = null;
			}
		}
		actors.clear();
	}

	public void setArea(RectBox area) {
		this.area = area;
		this.areaRipple = true;
	}

	public void setSplitAxis(float axis) {
		if (axis != this.splitAxis) {
			this.splitAxis = axis;
			this.areaRipple = true;
		}
	}

	public void setSplitPos(float pos) {
		if (pos != this.splitPos) {
			this.splitPos = pos;
			this.areaRipple = true;
		}

	}

	public float getSplitAxis() {
		return this.splitAxis;
	}

	public float getSplitPos() {
		return this.splitPos;
	}

	public RectBox getLeftArea() {
		return this.splitAxis == 0 ? new RectBox(this.area.getX(),
				this.area.getY(), this.splitPos - this.area.getX(),
				this.area.getHeight()) : new RectBox(this.area.getX(),
				this.area.getY(), this.area.getWidth(), this.splitPos
						- this.area.getY());
	}

	public RectBox getRightArea() {
		return this.splitAxis == 0 ? new RectBox(this.splitPos,
				this.area.getY(), this.area.getRight() - this.splitPos,
				this.area.getHeight()) : new RectBox(this.area.getX(),
				this.splitPos, this.area.getWidth(), this.area.getBottom()
						- this.splitPos);
	}

	public RectBox getArea() {
		return this.area;
	}

	private void resizeChildren() {
		if (this.left != null) {
			this.left.setArea(this.getLeftArea());
		}

		if (this.right != null) {
			this.right.setArea(this.getRightArea());
		}

	}

	public BSPCollisionNode getLeft() {
		if (this.areaRipple) {
			this.resizeChildren();
			this.areaRipple = false;
		}
		return this.left;
	}

	public BSPCollisionNode getRight() {
		if (this.areaRipple) {
			this.resizeChildren();
			this.areaRipple = false;
		}
		return this.right;
	}

	public BSPCollisionNode getParent() {
		return this.parent;
	}

	public void setParent(BSPCollisionNode parent) {
		this.parent = parent;
	}

	public int getChildSide(BSPCollisionNode child) {
		return this.left == child ? 0 : 1;
	}

	public void addActor(Actor actor) {
		this.actors.put(actor, new ActorNode(actor, this));
	}

	public boolean containsActor(Actor actor) {
		ActorNode anode = this.actors.get(actor);
		if (anode != null) {
			anode.mark();
			return true;
		} else {
			return false;
		}
	}

	public void actorRemoved(Actor actor) {
		this.actors.remove(actor);
	}

	public int numberActors() {
		return this.actors.size();
	}

	public boolean isEmpty() {
		return this.actors.isEmpty();
	}

	public Iterator getEntriesIterator() {
		return this.actors.entrySet().iterator();
	}

	public Iterator getActorsIterator() {
		return this.actors.keySet().iterator();
	}

	public List getActorsList() {
		return new ArrayList(this.actors.keySet());
	}

}
